core/src/apps/cardano/helpers/tx_builder_state_machine.py

```diff
@@ -0,0 +1,166 @@
+from micropython import const
+
+if False:
+    ActionType = int
+
+
+class TxBuilderStateMachine:
```

Comment:
I will have to take a closer look at this - but it does sound reasonable and doable.

Reply:
In the end I've inlined the calls to the hash builder and thus `tx_builder` and the state machine have been deleted. I tried implementing it via the generator but the inner structure of the generator was basically a duplicate of the branching in the `process_xxx` calls and also lacked type and other guarantees. It also worsened readability significantly, which could perhaps be improved in a few iterations, but IMHO it's not worth it. If you're interested you can take a look https://github.com/vacuumlabs/trezor-firmware/pull/67 - I'd be interested in hearing some alternatives/improvements to how it can written with a generator - perhaps I went the wrong way with the function yielding.

Reply:
wrt generator usage, what I had in mind was sort of the inverse of what you are trying at [`a5e01a1` (#67)](https://github.com/vacuumlabs/trezor-firmware/pull/67/commits/a5e01a15ab2ab4b15bfc7982928a813af2e8334f)

simplified pseudocode:
```
def build_tx_hash(params):
	hash_builder = HashBuilder()
	hash_builder.add(params.foo)
	hash_builder.add(params.bar)
	for _ in range(params.num_inputs):
		inp = yield
		assert isinstance(inp, CardanoInput)
		yield from build_tx_hash_input(hash_builder, inp)

def build_tx_hash_input(inp):
	hash_builder.add_collection("input")
	hash_builder.add(inp.amount)
	for _ in range(inp.metadata_len):
		meta = yield
		assert isinstance(meta, CardanoInputMeta)
		hash_builder.add(meta)

def sign_tx(msg):
	builder = build_tx_hash(msg)
	for inp in get_inputs():
		builder.send(inp)
		builder.send(inp.meta1)
		builder.send(inp.meta2)
	for out in get_outputs():
		builder.send(out)
```

it's much more readable, but it still has the properties that (a) it copies the layout of the `process_xxx` calls and (b) lacks type guarantees for callers.

---

i'm perfectly OK with the current version, and unless you have compelling reasons to go the generator route, i'd keep the inlined version